src/crimson/admin/osd_admin.cc

```diff
@@ -0,0 +1,318 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "crimson/admin/osd_admin.h"
+
+#include <iostream>
+
+#include <boost/algorithm/string.hpp>
+#include <fmt/format.h>
+#include <seastar/core/do_with.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/thread.hh>
+
+#include "common/config.h"
+#include "crimson/admin/admin_socket.h"
+#include "crimson/common/log.h"
+#include "crimson/osd/exceptions.h"
+#include "crimson/osd/osd.h"
+
+using crimson::osd::OSD;
+
+namespace {
+seastar::logger& logger()
+{
+  return crimson::get_logger(ceph_subsys_osd);
+}
+}  // namespace
+
+namespace crimson::admin {
+
+using crimson::common::local_conf;
+
+/**
+ * the hooks and states needed to handle OSD asok requests
+ */
+class OsdAdminImp {
+  friend class OsdAdmin;
+  friend class OsdAdminHookBase;
+
+  OSD* m_osd;
+
+  //  shared-ownership of the socket server itself, to guarantee its existence
+  //  until we have a chance to remove our registration:
+  AsokRegistrationRes m_socket_server;
+
+  /**
+   * Common code for all OSD admin hooks.
+   * Adds access to the owning OSD.
+   */
+  class OsdAdminHookBase : public AdminSocketHook {
+   protected:
+    explicit OsdAdminHookBase(OsdAdminImp& master) : m_osd_admin{ master } {}
+    struct tell_result_t {
+      int ret = 0;
+      string err;
+    };
+    /// the specific command implementation
+    virtual seastar::future<tell_result_t> tell(const cmdmap_t& cmdmap,
+						Formatter* f) const = 0;
+    seastar::future<bufferlist> call(std::string_view command,
+				     std::string_view format,
+				     const cmdmap_t& cmdmap) const final
+    {
+      unique_ptr<Formatter> f{Formatter::create(format, "json-pretty", "json-pretty")};
+      auto ret = seastar::do_with(std::move(f), [cmdmap, this] (auto& f) {
+	Formatter* formatter = f.get();
+	return tell(cmdmap, formatter).then([this, formatter](auto result) {
+	  bufferlist out;
+	  if (auto& [ret, err] = result; ret < 0) {
+	    out.append(fmt::format("ERROR: {}\n", cpp_strerror(ret)));
+	    out.append(err);
+	  } else {
+	    formatter->flush(out);
+	  }
+	  return seastar::make_ready_future<bufferlist>(std::move(out));
+        });
+      });
+      return ret;
+    }
```

Comment:
Another issue: I'd really like 'call()' here to be protected against tell() throwing or returning a failed future. First - because we want to be able to send the error code out, and second - as these commands arrive from elsewhere, this might be later opening an attack surface.

Reply:
> because we want to be able to send the error code out, 

agreed. that's why tell commands should take care of themselves by returning `tell_result_t`. `call()` should not interpret the exception or do the error handling.

> as these commands arrive from elsewhere, this might be later opening an attack surface.

could you be more specific on this?

Reply:
Yes. Will take this offline.